receiver/signalfxreceiver/receiver.go

```diff
@@ -0,0 +1,255 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfxreceiver
+
+import (
+	"compress/gzip"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/gorilla/mux"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/observability"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/receiver"
+	"github.com/open-telemetry/opentelemetry-collector/translator/conventions"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf"
+	"go.opencensus.io/trace"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultServerTimeout = 20 * time.Second
+
+	responseInvalidMethod      = "Only \"POST\" method is supported"
+	responseInvalidContentType = "\"Content-Type\" must be \"application/x-protobuf\""
+	responseInvalidEncoding    = "\"Content-Encoding\" must be \"gzip\" or empty"
+	responseErrGzipReader      = "Error on gzip body"
+	responseErrReadBody        = "Failed to read message body"
+	responseErrUnmarshalBody   = "Failed to unmarshal message body"
+	responseErrNextConsumer    = "Internal Server Error"
+)
+
+var (
+	errNilNextConsumer = errors.New("nil nextConsumer")
+	errEmptyEndpoint   = errors.New("empty endpoint")
+)
+
+// sfxReceiver implements the receiver.TraceReceiver for Zipkin Scribe protocol.
+type sfxReceiver struct {
+	sync.Mutex
+	logger       *zap.Logger
+	config       *Config
+	nextConsumer consumer.MetricsConsumer
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ receiver.MetricsReceiver = (*sfxReceiver)(nil)
+
+func (r *sfxReceiver) MetricsSource() string {
+	const metricsSource string = "SignalFx"
+	return metricsSource
+}
+
+// New creates the SignalFx receiver with the given configuration.
+func New(
+	logger *zap.Logger,
+	config Config,
+	nextConsumer consumer.MetricsConsumer,
+) (receiver.MetricsReceiver, error) {
+
+	if nextConsumer == nil {
+		return nil, errNilNextConsumer
+	}
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	// Handle config zero values.
+	if config.Name() == "" {
+		config.SetType(typeStr)
+		config.SetName(typeStr)
+	}
+
+	r := &sfxReceiver{
+		logger:       logger,
+		config:       &config,
+		nextConsumer: nextConsumer,
+		server: &http.Server{
+			Addr: config.Endpoint,
+			// TODO: Evaluate what properties should be configurable, for now
+			//		set some hard-coded values.
+			ReadHeaderTimeout: defaultServerTimeout,
+			WriteTimeout:      defaultServerTimeout,
+		},
+	}
+
+	mux := mux.NewRouter()
+	mux.HandleFunc("/v2/datapoint", r.handleReq)
+	r.server.Handler = mux
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the data received is set at creation time.
```

Comment:
This is an old conversation regarding where/when to set the next consumer: one option was to set it at start or in a separate method so it would become part of the interface, the other was to do it at creation time. In the end of the discussion people preferred to set it at creation time, this sentence tries to capture that. BTW, it is the same in the interface definition on core repo, perhaps worth an update there to.

Reply:
I see. So this refers to the fact the consumer is provided in `New` and `StartMetricsReception` is going to send to that consumer.

Perhaps rephrase it to make it clearer, I was a bit confused.